### PR TITLE
Generic process implementation

### PR DESCRIPTION
**Pull request by keyonjie**
This is a simple series to support generic process component as described in:
https://github.com/orgs/thesofproject/teams/sof-developers/discussions/32

UUID support is not added yet, but this is sufficient to unblock adding new process components like crossover, beamformer, ...

The FW corresponding PR is: https://github.com/thesofproject/sof/pull/2956

After this series is applied, all existed process components including mux/demux/eq/kpb/detect_test/selector/smart_amp/dcblock will be switched to this generic process type automatically.

The ABI minor will be bumped to 17 with this change.

### CODE REVIEW

**Comment by keyonjie** [comment]
@lgirdwood @mmaka1 @kv2019i I am still using 32bit subtype as the 16Bytes UUID dereference is not implemented yet inside the FW, if we decide to go with that, we can add incremental changes once it is implemented.

File: sound/soc/sof/topology.c [hunk 1]
```diff
@@ -2183,7 +2183,8 @@ static int sof_process_load(struct snd_soc_component *scomp, int index,
 	process->comp.hdr.size = ipc_size;
 	process->comp.hdr.cmd = SOF_IPC_GLB_TPLG_MSG | SOF_IPC_TPLG_COMP_NEW;
 	process->comp.id = swidget->comp_id;
-	process->comp.type = type;
+	process->comp.type = SOF_COMP_PROCESS;
+	process->comp.subtype = type;
```
Comment (kv2019i) [inline sound/soc/sof/topology.c]: @keyonjie Should this be conditional to FW version? If kernel is used with older firmware, this would break this case, right?

File: include/sound/sof/topology.h [hunk 2]
```diff
@@ -59,8 +59,7 @@ struct sof_ipc_comp {
 	uint32_t pipeline_id;
```
Comment (dbaluta) [inline include/sound/sof/topology.h:63]: s/flavour/favour in commit message?

File: include/sound/sof/topology.h [hunk 3]
```diff
@@ -39,6 +39,8 @@ enum sof_comp_type {
 	SOF_COMP_ASRC,		/**< Asynchronous sample rate converter */
 	SOF_COMP_DCBLOCK,
```
Comment (dbaluta) [inline include/sound/sof/topology.h:40]: @keyonjie change per se looks good to me, after we address the backward compatibility issues. Is there any documentation about how do we keep backward compatibility?

E.g:
- newer kernels should work with older FW if the major number is the same
- newer FW should work with older kernels if the major number is the same

File: include/sound/sof/topology.h [hunk 4]
```diff
@@ -39,6 +39,8 @@ enum sof_comp_type {
...
 	SOF_COMP_SMART_AMP,             /**< smart amplifier component */
+
+	SOF_COMP_PROCESS = 1000,	/**< generic process component */
 	/* keep FILEREAD/FILEWRITE as the last ones */
 	SOF_COMP_FILEREAD = 10000,	/**< host test based file IO */
 	SOF_COMP_FILEWRITE = 10001,	/**< host test based file IO */
```
Comment (dbaluta) [inline include/sound/sof/topology.h:50]: @keyonjie can we enhance the commit message to specify WHY this change is needed. I'm not really sure but I think it has something to do with the effort of adding new components which are simpler enough to be abstracted to just ONE processing function.

File: sound/soc/sof/topology.c (hunk 1 above)
Comment (keyonjie) [reply sound/soc/sof/topology.c]: Good point, let me add a check here.

File: include/sound/sof/topology.h (hunk 2 above)
Comment (keyonjie) [reply include/sound/sof/topology.h:63]: oops, thanks.

File: include/sound/sof/topology.h (hunk 3 above)
Comment (keyonjie) [reply include/sound/sof/topology.h:40]: We don't have documentation about this, per my understanding, @plbossart mentioned that we should keep backward compatible as the 1st item you mentioned as user might keep updating kernel while firmware usually keep intact:
"newer kernels should work with older FW if the major number is the same"

we have a kconfig item CONFIG_SND_SOC_SOF_STRICT_ABI_CHECKS and newer FW will be rejected when the item is selected.

File: include/sound/sof/topology.h (hunk 4 above)
Comment (keyonjie) [reply include/sound/sof/topology.h:50]: Maybe I can add some message from the PR to one of the commit for this. 
@dbaluta the corresponding FW PR is provided, we might have more refinement (e.g. share common codes among different process component) later but now all existed process components should work with these 2 PRs already.

Comment (dbaluta) [reply include/sound/sof/topology.h:50]: Ok, sounds good!

File: sound/soc/sof/topology.c [hunk 5]
```diff
@@ -2183,7 +2183,13 @@ static int sof_process_load(struct snd_soc_component *scomp, int index,
 	process->comp.hdr.size = ipc_size;
 	process->comp.hdr.cmd = SOF_IPC_GLB_TPLG_MSG | SOF_IPC_TPLG_COMP_NEW;
 	process->comp.id = swidget->comp_id;
-	process->comp.type = type;
+	if (sdev->fw_ready.version.abi_version >= SOF_ABI_VER(3, 17, 0)) {
+		/* use generic process component if supported */
+		process->comp.type = SOF_COMP_PROCESS;
+		process->comp.subtype = type;
```
Comment (plbossart) [inline sound/soc/sof/topology.c:2189]: how is this subtype defined? and how does this help make the difference between e.g. an encoder and an EQ? You are missing one layer of description IMHO.

Comment (plbossart) [reply sound/soc/sof/topology.c:2189]: also how does this relate to PR #2090? Please don't provide disjoint PRs for the same topic, it's not nice for reviewers.

Comment (keyonjie) [reply sound/soc/sof/topology.c:2189]: > how is this subtype defined? and how does this help make the difference between e.g. an encoder and an EQ? You are missing one layer of description IMHO.

At the moment I just use the component type as the subtype for the process components, as the EQ already has its component type. 

Thanks for your feedback Pierre. We are actually not quite aligned about what direction should we go (see the discussion here https://github.com/thesofproject/sof/pull/2956#issuecomment-631833396).

You are right that once we have decided to go with subtype, I need to add definition for the subtype (e.g. make the subtype another enum, move all existed process component types' definition into this enum, and this new defined enum should be kept in the header file shared by FW and topology only, the linux driver won't know about this).

Comment (keyonjie) [reply sound/soc/sof/topology.c:2189]: > also how does this relate to PR #2090? Please don't provide disjoint PRs for the same topic, it's not nice for reviewers.

Sorry for the confusion, yes, these 2 PRs are disjoint, they are 2 different solutions as described here: https://github.com/thesofproject/sof/pull/2956#issuecomment-631833396

Comment (plbossart) [reply sound/soc/sof/topology.c:2189]: Man, you are really not making it easy for reviewers to look into your work...Please try and put yourself in the shoes of someone not familiar with your ideas and without context on another email thread.

As shown with PR #2138 I had to do the integration work myself to figure out what to do with your patches and what you meant in commit messages. Try and explain more what you are trying to fix, it'll help have your patches merged quicker and with fewer review cycles.

Comment (keyonjie) [reply sound/soc/sof/topology.c:2189]: Thanks for your suggestion here, I don't know how to say, I would forgive myself here, after joined the related discussion meetings, created 2 discussion topics, verified and provided 2 solutions for it, maybe the commit messages are still not good, but had tried my best.

File: sound/soc/sof/topology.c [hunk 6]
```diff
@@ -2183,7 +2183,13 @@ static int sof_process_load(struct snd_soc_component *scomp, int index,
...
+	} else {
+		process->comp.type = SOF_COMP_PROCESS;
```
Comment (lyakh) [inline sound/soc/sof/topology.c:2191]: shouldn't this be `= type`?

Comment (keyonjie) [reply sound/soc/sof/topology.c:2191]: yep, good catch, thanks.

File: sound/soc/sof/topology.c (hunk 5 above)
Comment (kv2019i) [reply sound/soc/sof/topology.c:2189]: I'll also chime in that this has been exceptionally confusing case, but for me @keyonjie 's PR have actually helper to clarify the problem. It seems the SOF discussions pages (like https://github.com/orgs/thesofproject/teams/sof-developers/discussions/32 ) work fairly well to brainstorm ideas, but we don't really seem to get actual decisions done there. To force a decision, submitting a PR works better (like @keyonjie you did here), but when you have issues that affect both driver and FW arch, this approach does not seem to work very well as the discussion threads get distributed across multiple PRs.

Not sure how to improve this. One option is drive to conclusion with a patch to sof-docs that documents the proposed design.

File: include/sound/sof/topology.h [hunk 7]
```diff
@@ -39,6 +39,12 @@ enum sof_comp_type {
 	SOF_COMP_ASRC,		/**< Asynchronous sample rate converter */
 	SOF_COMP_DCBLOCK,
 	SOF_COMP_SMART_AMP,             /**< smart amplifier component */
+
+	/*
+	 * No more _COMP_ types to be added.
+	 * Use SOF_COMP_PROCESS from now on.
```
Comment (kv2019i) [inline include/sound/sof/topology.h:45]: @keyonjie @lgirdwood  Shouldn't this be "No more _COMP_ types should be added for effects" (snd_soc_dapm_effect/sof_ipc_comp_process)?

If want to have main type that defines the ALSa controls and IPC msg, and flavor which can vary, we can't put everything under same component id right? The above comment would mean our current list of main types is exhaustive and it will never be extended. This doesn't make sense. If we add a new component class, it should be identified by a unique component type.

File: include/sound/sof/topology.h [hunk 8]
```diff
@@ -39,6 +39,12 @@ enum sof_comp_type {
 	SOF_COMP_ASRC,		/**< Asynchronous sample rate converter */
```
Comment (kv2019i) [inline include/sound/sof/topology.h:39]: @keyonjie The discussions page is not public for all and we didn't really conclude anything there, so I would not refer to it from the commit message, but rather summarize the design here. I'd also not refer to the FW PR in the commit message (we have not done that traditionally in the kernel commit messages -> rather have these in the kernel PR description). I'd also drop the paragraph on "After this series is applied.."

File: include/sound/sof/topology.h [hunk 9]
```diff
@@ -63,8 +63,7 @@ struct sof_ipc_comp {
 	uint32_t pipeline_id;
 	uint32_t core;
 
-	/* reserved for future use */
-	uint32_t reserved[1];
+	uint32_t subtype;	/**< flavour for generic component type */
 } __packed;
```
Comment (kv2019i) [inline include/sound/sof/topology.h:67]: The FW discussion ( https://github.com/thesofproject/sof/pull/2956 ) seems to be moving to use UUIDs directly, so let's target that.

File: include/sound/sof/topology.h (hunk 7 above)
Comment (keyonjie) [reply include/sound/sof/topology.h:45]: Yes, let me change it to "No more _COMP_ types should be added for effects/process".

File: include/sound/sof/topology.h (hunk 8 above)
Comment (keyonjie) [reply include/sound/sof/topology.h:39]: Good points, thank you.

File: include/sound/sof/topology.h (hunk 9 above)
Comment (keyonjie) [reply include/sound/sof/topology.h:67]: Okay.

File: sound/soc/sof/topology.c (hunk 5 above)
Comment (keyonjie) [reply sound/soc/sof/topology.c:2189]: Thank you @kv2019i for the constructive suggestion, let's try to finalize the solution in the next tech sync meeting, and then I will close PRs for one of the two solutions.

**Comment by keyonjie** [comment]
The aligned solution is stated here https://github.com/orgs/thesofproject/teams/sof-developers/discussions/36/comments/22. According to it, we will need to switch to use UUID for all component types in one shot, with a major ABI bump for it. So no need for this generic process component anymore if that is implemented, let me close PRs and work on the new solution.

File: sound/soc/sof/topology.c (hunk 5 above)
Comment (plbossart) [reply sound/soc/sof/topology.c:2189]: having an sof-docs PR is really the only way to close on this.

**Comment by keyonjie** [comment]
@kv2019i @plbossart Let me close this PR and use the only one PR for this UUID implementation on kernel side: #2090 to avoid confusion.
